src/mono/System.Private.CoreLib/src/System/Threading/Thread.Mono.cs

```diff
@@ -206,15 +205,15 @@ private void Initialize()
         {
             InitInternal(this);
 #if TARGET_UNIX || TARGET_BROWSER
-            _waitInfo = new WaitSubsystem.ThreadWaitInfo(this);
+            EnsureWaitInfo();
 #endif
         }
 
 #if TARGET_UNIX || TARGET_BROWSER
-        private void EnsureWaitInfo()
+        [MemberNotNull(nameof(_waitInfo))]
+        private WaitSubsystem.ThreadWaitInfo EnsureWaitInfo()
         {
-            if (_waitInfo == null)
-                _waitInfo = new WaitSubsystem.ThreadWaitInfo(this);
+            return LazyInitializer.EnsureInitialized(ref _waitInfo, () => new WaitSubsystem.ThreadWaitInfo(this));
         }
```

Comment:
This _might_ have perf problems, some hand written double locked code might be better.

Reply:
I looked at the implementation and fast path of EnsureInitialized and it does a volatile read and a branch:

```
public static T EnsureInitialized<T>([NotNull] ref T? target, Func<T> valueFactory) where T : class =>
            Volatile.Read(ref target!) ?? EnsureInitializedCore(ref target, valueFactory);
```

Only way to make faster is to do regular read, but then we are playing on the boundary of correctness since it will depend on compiler/hardware optimization, but since its a data dependency, I believe it will still be doing the right thing, at least on the CPU arch currently supported by runtime, but still, using volatile read will be more correct and using EnsureInitialized should be inlined into caller, so I believe this should be optimized into doing the same handwritten code, or did you have something else in mind?

Reply:
Also note that I only found references to the .WaitInfo property in functions that are quite expensive to begin with, like Sleep, NewMutex, Wait, SignalAndWait, Interrupt and RelinquishOwnership (called from SignalMutex).

Reply:
The expensive part is the lamba allocation. The lambda `() => new WaitSubsystem.ThreadWaitInfo(this)` is going to run every time that `EnsureWaitInfo` is called. It is useless work 99+% of the time.

> using EnsureInitialized should be inlined into caller

I doubt it that EnsureInitialized be inlined with this change. The code to allocate lamda closure is pretty big.

Reply:
OK, well if that prevents inlining then I guess we need to manual inline the expression.

Reply:
Did a quick check on x64 (using CoreClr) after change, so the manual inlined code will be inlined into caller, looking like this:

lea         rax,[rcx+8]  
mov       rax,qword ptr [rax]  
test        rax,rax  
jne         00007FFAD41908C1  
call        00007FFAD41904D0  
mov      eax,3B9ACA00h

Since this is x64, the load volatile will just be a regular lea/move and then a conditional branch, all inlined into caller.